File: apps/driver-android/src/main/kotlin/lk/mageride/driver/earnings/EarningsScreen.kt
package lk.mageride.driver.earnings

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.ArrowBack
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.driver.R
import lk.mageride.driver.jobs.ScheduleLabels
import lk.mageride.driver.ui.MoneyFormat
import lk.mageride.driver.ui.component.DashboardBanner
import lk.mageride.driver.ui.component.SectionLabel
import lk.mageride.driver.ui.theme.MageRideTheme
import lk.mageride.shared.data.models.query.SessionEarning
import org.koin.androidx.compose.koinViewModel
import kotlin.time.ExperimentalTime

/**
 * **SCR-DA-020 · the earnings dashboard** (US-9.22, US-8.8).
 *
 * The wireframe, top to bottom: a `‹` app bar, the **Today / Week / Month** tab bar, the period's
 * net in display type, the bar chart, and the card that shows what the net is made of — fares
 * received, the daily fee deducted in `error`, and the bold **Net** line. The per-trip rows D2'
 * lists in its component table follow underneath.
 *
 * Read [EarningsViewModel] before adding a figure: the summary is query-svc's and is printed as
 * sent, and the payment-method split the deliverable asks for has no read behind it.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun EarningsScreen(onBack: () -> Unit, modifier: Modifier = Modifier) {
    val viewModel: EarningsViewModel = koinViewModel()
    val state by viewModel.state.collectAsStateWithLifecycle()

    Scaffold(
        modifier = modifier.fillMaxSize(),
        topBar = {
            TopAppBar(
                title = { Text(text = stringResource(R.string.earnings_title)) },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(
                            imageVector = Icons.AutoMirrored.Outlined.ArrowBack,
                            contentDescription = stringResource(R.string.action_back),
                        )
                    }
                },
            )
        },
    ) { insets ->
        Column(modifier = Modifier.padding(insets).fillMaxSize()) {
            EarningsPeriodTabs(selected = state.period, onSelect = viewModel::select)

            state.error?.let { message ->
                DashboardBanner(text = stringResource(message), accent = MaterialTheme.colorScheme.error)
            }

            LazyColumn(
                modifier = Modifier.fillMaxSize(),
                contentPadding = PaddingValues(MageRideTheme.spacing.md),
                verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
            ) {
                item { PeriodTotal(state = state) }
                item { EarningsChart(buckets = state.buckets) }
                item { BreakdownCard(state = state) }

                if (state.isEmptyPeriod) {
                    item { EmptyPeriod() }
                } else {
                    item { SectionLabel(text = stringResource(R.string.earnings_trips)) }
                    items(items = state.trips, key = { it.tripId }) { trip -> TripRow(trip = trip) }
                }
            }
        }
    }
}

/** *"Today's earnings — Rs 3,180"*. The label names the selected window, not the day. */
@Composable
private fun PeriodTotal(state: EarningsState, modifier: Modifier = Modifier) {
    Column(
        modifier = modifier.fillMaxWidth(),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Text(
            text = stringResource(R.string.earnings_net_for, stringResource(state.period.labelRes())),
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        Text(
            text = MoneyFormat.rupees(state.netMinor),
            style = MaterialTheme.typography.displaySmall,
            color = MaterialTheme.colorScheme.onSurface,
        )
    }
}

/**
 * What the net is made of.
 *
 * The wireframe prints three lines — fares received, the daily fee, the net. Penalties (D5' §7.1)
 * and tips (E-10) are two more fields `EarningsSummary` carries, and each is shown **only when the
 * server sent a non-zero one**: a driver who has never been penalised should not have a penalty line
 * on their dashboard, and a driver who has must not have it folded silently into the net.
 */
@Composable
private fun BreakdownCard(state: EarningsState, modifier: Modifier = Modifier) {
    Card(
        modifier = modifier.fillMaxWidth(),
        shape = RoundedCornerShape(MageRideTheme.radius.md),
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surfaceVariant),
        elevation = CardDefaults.cardElevation(defaultElevation = MageRideTheme.elevation.level0),
    ) {
        Column(
            modifier = Modifier.padding(MageRideTheme.spacing.sm),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xxs),
        ) {
            MoneyRow(
                label = stringResource(R.string.earnings_fares_received, state.tripCount),
                value = MoneyFormat.rupees(state.grossMinor),
            )

            if (state.tipMinor > 0) {
                MoneyRow(
                    label = stringResource(R.string.earnings_tips),
                    value = MoneyFormat.rupees(state.tipMinor),
                )
            }

            if (state.dailyFeeMinor > 0) {
                MoneyRow(
                    label = stringResource(R.string.earnings_daily_fee),
                    value = MoneyFormat.rupees(-state.dailyFeeMinor),
                    accent = MaterialTheme.colorScheme.error,
                )
            }

            if (state.penaltyMinor > 0) {
                MoneyRow(
                    label = stringResource(R.string.earnings_penalties),
                    value = MoneyFormat.rupees(-state.penaltyMinor),
                    accent = MaterialTheme.colorScheme.error,
                )
            }

            HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant)

            MoneyRow(
                label = stringResource(R.string.earnings_net),
                value = MoneyFormat.rupees(state.netMinor),
                weight = FontWeight.Bold,
            )
        }
    }
}

/** The wireframe's `kv` — a label on the left and an amount on the right. */
@Composable
private fun MoneyRow(
    label: String,
    value: String,
    modifier: Modifier = Modifier,
    accent: Color? = null,
    weight: FontWeight? = null,
) {
    Row(
        modifier = modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            fontWeight = weight,
            modifier = Modifier.weight(1f),
        )
        Text(
            text = value,
            style = MaterialTheme.typography.bodyMedium,
            color = accent ?: MaterialTheme.colorScheme.onSurface,
            fontWeight = weight,
        )
    }
}

/**
 * One completed trip (US-8.8).
 *
 * The time is the trip's **Colombo** end time, for the same reason the chart's buckets are — the
 * period the card totalled is a Colombo one.
 */
@OptIn(ExperimentalTime::class)
@Composable
private fun TripRow(trip: SessionEarning, modifier: Modifier = Modifier) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .padding(vertical = MageRideTheme.spacing.xxs),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
    ) {
        Text(
            text = ScheduleLabels.time(trip.endedAt),
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        Text(
            text = MoneyFormat.rupees(trip.grossMinor),
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.weight(1f),
            textAlign = TextAlign.End,
        )
        Text(
            text = MoneyFormat.rupees(trip.netMinor),
            style = MaterialTheme.typography.titleMedium,
            color = MaterialTheme.colorScheme.onSurface,
        )
    }
}

/** D2' §SCR-DA-020's *"empty period"*. */
@Composable
private fun EmptyPeriod(modifier: Modifier = Modifier) {
    Text(
        text = stringResource(R.string.earnings_empty),
        style = MaterialTheme.typography.bodyLarge,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        textAlign = TextAlign.Center,
        modifier = modifier
            .fillMaxWidth()
            .padding(MageRideTheme.spacing.lg),
    )
}

File: apps/driver-ios/DriverApp/Earnings/EarningsScreen.swift
import MageRideShared
import SwiftUI

/// **SCR-DI-020 · the earnings dashboard** (US-9.22, US-8.8).
///
/// The wireframe, top to bottom: the large title *"Earnings"*, the **Today / Week / Month** segmented
/// control, the period's net in display type, the bar chart, and the card that shows what the net is
/// made of — fares received, the daily fee deducted in `error`, and the bold **Net** line. The
/// per-trip rows D2' lists in its component table follow underneath.
///
/// Read ``EarningsModel`` before adding a figure: the summary is query-svc's and is printed as sent,
/// and the payment-method split the deliverable asks for has no read behind it.
///
/// Opened from SCR-DI-010's *"Today: 4 trips · Rs 3,180"* line — which **is** this screen's Today
/// figure, the same `EarningsSummary.netMinor` through the same endpoint — and C088 already wired it.
///
/// `@MainActor` on the whole view, not on its initialiser — see ``ProfileSetupScreen`` for why.
@MainActor
struct EarningsScreen: View {

    @StateObject private var model: EarningsModel

    init(model: @autoclosure @escaping () -> EarningsModel) {
        _model = StateObject(wrappedValue: model())
    }

    var body: some View {
        ScrollView {
            VStack(spacing: MageRideSpacing.sm) {
                EarningsPeriodTabs(selected: model.state.period) { period in
                    Task { await model.select(period) }
                }

                if let errorKey = model.state.errorKey {
                    DashboardBanner(text: errorKey.localised, accent: MageRideColor.error)
                        .onTapGesture(perform: model.dismissError)
                }

                periodTotal
                EarningsChart(buckets: model.state.buckets)
                breakdownCard

                if model.state.isEmptyPeriod {
                    Text(key: "earnings_empty")
                        .mageFont(.body)
                        .foregroundStyle(MageRideColor.onSurfaceVariant)
                        .multilineTextAlignment(.center)
                        .padding(MageRideSpacing.lg)
                        .frame(maxWidth: .infinity)
                } else {
                    SectionLabel(key: "earnings_trips")
                    ForEach(model.state.trips, id: \.tripId) { trip in
                        TripRow(trip: trip)
                    }
                }
            }
            .padding(MageRideSpacing.md)
        }
        .frame(maxWidth: .infinity)
        .background(MageRideColor.background)
        .navigationTitle(Text(key: "earnings_title"))
        .navigationBarTitleDisplayMode(.large)
        .task { await model.refresh() }
    }

    /// *"Today — what you keep · Rs 3,180"*. The label names the selected window, not the day.
    private var periodTotal: some View {
        VStack(spacing: 1) {
            Text("earnings_net_for".localisedFormat(model.state.period.labelKey.localised))
                .mageFont(.caption)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
            Text(MoneyFormat.rupees(model.state.netMinor))
                .mageFont(.display)
                .foregroundStyle(MageRideColor.onSurface)
        }
        .frame(maxWidth: .infinity)
        .accessibilityElement(children: .combine)
    }

    /// What the net is made of.
    ///
    /// The wireframe prints three lines — fares received, the daily fee, the net. Penalties
    /// (D5' §7.1) and tips (E-10) are two more fields `EarningsSummary` carries, and each is shown
    /// **only when the server sent a non-zero one**: a driver who has never been penalised should not
    /// have a penalty line on their dashboard, and a driver who has must not have it folded silently
    /// into the net.
    private var breakdownCard: some View {
        VStack(spacing: MageRideSpacing.xxs) {
            MoneyRow(
                label: "earnings_fares_received".localisedFormat(model.state.tripCount),
                value: MoneyFormat.rupees(model.state.grossMinor)
            )

            if model.state.tipMinor > 0 {
                MoneyRow(label: "earnings_tips".localised, value: MoneyFormat.rupees(model.state.tipMinor))
            }

            if model.state.dailyFeeMinor > 0 {
                MoneyRow(
                    label: "earnings_daily_fee".localised,
                    value: MoneyFormat.rupees(-model.state.dailyFeeMinor),
                    accent: MageRideColor.error
                )
            }

            if model.state.penaltyMinor > 0 {
                MoneyRow(
                    label: "earnings_penalties".localised,
                    value: MoneyFormat.rupees(-model.state.penaltyMinor),
                    accent: MageRideColor.error
                )
            }

            Divider().overlay(MageRideColor.outlineVariant)

            MoneyRow(
                label: "earnings_net".localised,
                value: MoneyFormat.rupees(model.state.netMinor),
                isEmphasised: true
            )
        }
        .padding(MageRideSpacing.sm)
        .background(
            MageRideColor.surfaceVariant,
            in: RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous)
        )
    }
}

/// The wireframe's `kv` — a label on the left and an amount on the right.
private struct MoneyRow: View {

    let label: String
    let value: String
    var accent: Color?
    var isEmphasised = false

    var body: some View {
        HStack(alignment: .firstTextBaseline, spacing: MageRideSpacing.xs) {
            Text(label)
                .mageFont(isEmphasised ? .bodyEmphasis : .bodySmall)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
                .frame(maxWidth: .infinity, alignment: .leading)
            Text(value)
                .mageFont(isEmphasised ? .bodyEmphasis : .bodySmall)
                .foregroundStyle(accent ?? MageRideColor.onSurface)
        }
        .accessibilityElement(children: .combine)
    }
}

/// One completed trip (US-8.8).
///
/// The time is the trip's **Colombo** end time, for the same reason the chart's buckets are — the
/// period the card totalled is a Colombo one.
private struct TripRow: View {

    let trip: SessionEarning

    var body: some View {
        HStack(alignment: .firstTextBaseline, spacing: MageRideSpacing.xs) {
            Text(ScheduleLabels.time(trip.endedAt))
                .mageFont(.bodySmall)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
            Text(MoneyFormat.rupees(trip.grossMinor))
                .mageFont(.bodySmall)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
                .frame(maxWidth: .infinity, alignment: .trailing)
            Text(MoneyFormat.rupees(trip.netMinor))
                .mageFont(.title)
                .foregroundStyle(MageRideColor.onSurface)
        }
        .padding(.vertical, MageRideSpacing.xxs)
        .accessibilityElement(children: .combine)
    }
}
